src/side-loaded/basic-sideload/verifier-zkapp.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  SmartContract,
+  state,
+  State,
+  method,
+  Field,
+  PublicKey,
+  VerificationKey,
+  DeployArgs,
+  Permissions,
+} from 'o1js';
+
+import {
+  EcdsaSideloadedProof,
+  KeccakSideloadedProof,
+} from './sideloaded-zkprograms.js';
+
+export { MyFavoriteZkApp };
+
+class MyFavoriteZkApp extends SmartContract {
+  @state(PublicKey) contractOwner = State<PublicKey>();
+  @state(Field) protectedValue = State<Field>();
+  // Verification key hashes for side-loaded proofs
+  @state(Field) ecdsaVerificationKeyHash = State<Field>();
+  @state(Field) keccakVerificationKeyHash = State<Field>();
+
+  async deploy(args: DeployArgs & { owner: PublicKey }) {
+    await super.deploy(args);
+
+    // Initialize contract state
+    this.protectedValue.set(Field(333));
+    this.contractOwner.set(args.owner);
+
+    // Set default permissions
+    this.account.permissions.set({
+      ...Permissions.default(),
+    });
+  }
+
+  /**
+   * Updates the protected value after verifying both ECDSA and Keccak proofs
+   */
+  @method async updateValue(
+    newProtectedValue: Field,
+    ecdsaProof: EcdsaSideloadedProof,
+    keccakProof: KeccakSideloadedProof,
+    ecdsaVerificationKey: VerificationKey,
+    keccakVerificationKey: VerificationKey
+  ) {
+    // Verify ECDSA verification key against stored hash
+    const storedEcdsaKeyHash =
+      this.ecdsaVerificationKeyHash.getAndRequireEquals();
+    ecdsaVerificationKey.hash.assertEquals(storedEcdsaKeyHash);
+
+    // Verify Keccak verification key against stored hash
+    const storedKeccakKeyHash =
+      this.keccakVerificationKeyHash.getAndRequireEquals();
+    keccakVerificationKey.hash.assertEquals(storedKeccakKeyHash);
+
+    // Verify ECDSA proof
+    ecdsaProof.verify(ecdsaVerificationKey);
+    ecdsaProof.publicOutput.isValid.assertTrue(
+      'ECDSA signature verification failed'
+    );
+
+    // Get ECDSA proof's public key from output
+    const signerPublicKey = ecdsaProof.publicOutput.publicKey;
+    // TODO: You can use the signer's public key here for additional checks
+    // For example: verify if the signer is authorized, maintain a list of trusted signers, etc.
+
+    // Verify Keccak proof
+    keccakProof.verify(keccakVerificationKey);
+    keccakProof.publicOutput.isValid.assertTrue(
+      'Keccak hash verification failed'
+    );
+
+    // Get Keccak proof's expected hash from input
+    const expectedKeccakHash = keccakProof.publicInput.hash;
+    // TODO: You can use the hash here for additional checks
+    // For example: verify if the hash matches some expected value, store it for future reference, etc.
+
+    // Update the protected value after all verifications pass
+    this.protectedValue.set(newProtectedValue);
+  }
+
+  /**
+   * Retrieves the current protected value
+   */
+  @method async getProtectedValue(): Promise<void> {
```

Comment:
Are you using this method anywhere?

Reply:
It’s not currently used anywhere in the example, but I included it because it might be useful when experimenting with the code.